memory_util.h

```diff
@@ -0,0 +1,298 @@
+#pragma once
+
+#include <cassert>
+#include <cstring>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <vector>
+
```

Comment:
Do we need all the functions in this file? It seems a little bit overkill. Try to keep the minimal version.